scio-core/src/main/java/com/spotify/scio/transforms/BatchDoFn.java

```diff
@@ -0,0 +1,120 @@
+package com.spotify.scio.transforms;
+
+import org.apache.beam.sdk.transforms.DoFn;
+import org.apache.beam.sdk.transforms.SerializableFunction;
+import org.apache.beam.sdk.transforms.windowing.BoundedWindow;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.*;
+
+/**
+ * Batches input into a desired batch size.
+ *
+ * <p>Elements are buffered until there are enough elements for a batch,
+ * at which point they are emitted to the output PCollection
+ *
+ * <p>Windows are preserved (batches contain elements from the same window).
+ * Batches are not spanning over bundles. Once a bundle is finished, the batch is emitted even if not full.
+ * This function can only batch 10 parallel windows. If new element comes from an 11th window,
+ * the bigger batch will be emitted to give room for this new element.
+ *
+ * @param <InputT>
+ */
+public class BatchDoFn<InputT> extends DoFn<InputT, Iterable<InputT>> {
+
+    private static final Logger LOG = LoggerFactory.getLogger(BatchDoFn.class);
+
+    private static final int MAX_LIVE_WINDOWS = 10;
```

Comment:
someone would ask this anyway :) why 10 and why not making it the default value instead of constant?

Reply:
Shamelessly stole beam's python implementation. the constant is defined [here](https://github.com/apache/beam/blob/00bbc72e005d741f632bb171b36d470f6544be55/sdks/python/apache_beam/transforms/util.py#L566)

Reply:
hehe so maybe we need to do a better job :))

Reply:
Do you think we should also expose this in the scala API, or we only give it as convenience if a user has a problem with the default, he could use the `BatchDoFn` with a custom param ?

Reply:
I don't see why not exposing it? with a default value so they don't have to worry about it if they don't want?